packages/upload-client/src/sharding.js

```diff
@@ -45,10 +55,29 @@ export class ShardingStream extends TransformStream {
         }
 
         const rootBlock = shard.at(-1)
-        if (rootBlock != null) {
-          controller.enqueue(
-            await encode(shard, options.rootCID ?? rootBlock.cid)
-          )
+        if (rootBlock == null) return
+
+        const rootCID = options.rootCID ?? rootBlock.cid
+        const headerLength = headerEncodingLength(rootCID)
+        // does the shard with the CAR header that _includes_ a root CID
+        // exceed the shard size?
+        if (headerLength + shardBlockLength > shardSize) {
+          const overage = headerLength + shardBlockLength - shardSize
+          const lastShard = []
+          let lastShardBlockLength = 0
+          while (lastShardBlockLength < overage) {
+            // need at least 1 block in original shard
+            if (shard.length < 2)
+              throw new Error(`block exceeds shard size: ${shard.at(-1)?.cid}`)
+            const block = shard[shard.length - 1]
+            shard.pop()
+            lastShard.unshift(block)
+            lastShardBlockLength += blockEncodingLength(block)
```

Comment:
Nit: I was very confused when I read this, comment said `1` block but then we had `< 2`, I think changing order of operations would make it easier to follow

```suggestion
            const block = shard[shard.length - 1]
            shard.pop()
            lastShard.unshift(block)
            lastShardBlockLength += blockEncodingLength(block)
            
            // need at least 1 block in original shard
            if (shard.length < 1)
              throw new Error(`block exceeds shard size: ${shard.at(-1)?.cid}`)
```

I also think that error message is somewhat misleading, it's combination of the block and it's CID what exceeds the block size limit, maybe message can be improved here.